### ISSUE TITLE
Rate limit iops per key

### ISSUE THREAD

**Issue by dorlaor** [scylladb/scylla#4703]
In a discussion around a customer issue where a single key overloads
a shard and ruins the 99% for all of the requests on that shard, we brought up today an option to have a rate limit per key (across the entire keyspace) to 'solve' such a case. 

The solution for example, all keys will be limited to 10k iops each.
One will be able to change it using alter keyspace/table.

Makes sense?

It's nice to do that per individual key for debug reasons too.

Avi suggests to use the top-partition infrastructure.

**Comment by dorlaor** [comment]
From @glst


If we try to be generic about this, we've seen many problems before in which the real issue was that a particular key had a lot of rows. A couple of requests from those keys are able to saturate the shard.

A better metric would be amount of rows read per key, and not amount of requests per key.

**Comment by tzach** [comment]
Row number can be misleading as well, as each row can be small or big.

**Comment by avikivity** [comment]
A hot partition can have zero rows and still dominate traffic (for both reads and writes). I don't think rows are the correct metric.

A request that reads many rows can be more expensive than a request that reads a single row, but it can also be less expensive.

**Comment by glommer** [comment]
A hot partition cannot have zero rows. At least not if we count a request
to the key without any row as a row.

To avoid confusion let's use the term request: a read for a key without any
rows is a request.

A read that reads two rows counts as two requests.

A request to read a larger row won't be that much more expensive in the
cpu, but it is a fair comment: at some level it will be more expensive.

The bottom line is: can we provide something that will encompass these
differences and work well with many cases, instead of only counting how
many select statements a partition got, that will only work well for simple
schemas ?

I think we can. Maybe we are can do a bandwidth limit, with a penalty for
metadata heavy partitions where the rows are too small since that is harder
on the cpu.

**Comment by glommer** [comment]
Maybe the best way to combine those is two different limits: a rows
returned per second limit (including the key) and a bandwidth limit in
bytes per second.

Easy to understand and easy for the user to set. And it convers about
everything we need

**Comment by avikivity** [comment]
A row is a row. Don't define it to be something else. 

A request has a baseline cost, a per row cost, and a pet byte cost. We can have some metric that combines them (row = 0.1 request) but ultimately the requests dominate.

**Comment by tzach** [comment]
Maybe a Token Bucket per partition is a good fit here, allowing for short bursts above the hard limit 
(or as a v2?)

**Comment by tzach** [comment]
Instead of a unifying metric to capture all "hot" use case, we can start with a simple rate limit, on the number of requests, add more metric later, and maybe combine them eventually.
A trivial reads request counter can mitigate 80% of the hot partition use cases, reducing its impact on the core and the system.

**Comment by dorlaor** [comment]
+1 for promoting the priority, while it can't rescue the hot row, it will
rescue the rest
of the accesses.

**Comment by slivne** [comment]
@avikivity - I think you had some suggestion around this (and also CQL syntax to provide the limits) - yet I maybe mistaken.

**Comment by avikivity** [comment]
To implement we can reuse `toppartitions_item_listener` but instead of counting, throw.

We'd need to adjust to decay the counts over time so we measure over a sliding window.

CQL:

```sql
CREATE TABLE tab ( ... )
WITH PER PARTITION RATE LIMIT = {
    'max_writes_per_second': 150, 
    'max_reads_per_second': 400,
    'time_window_seconds': 30
};
```

**Comment by avikivity** [comment]
By reused, I mean a new class that is similar, not hack it to do two things.

**Comment by tzach** [comment]
The problem is its hard to pick numbers which make sense
It would be good to have some reasonable defaults, but its impossible in absolute values.

**Comment by dorlaor** [comment]
Tzach, usually these limits will be placed to solve an existing
problem, and I see ALTER_TABLE used more than create table.
So defaults matter less.

Can we define an error code to return, so the client won't need to wait
for a time out?

**Comment by tzach** [comment]
> usually these limits will be placed to solve an existing
problem

Why? it make sense no one partition to take more than 1/x of the load of the table
Setting it as default will stop many issue from happening in the first place

~X can be the number of cores in the system.

**Comment by avikivity** [comment]
I don't see how you can calculate this. We know the number of shards, but we don't know anything else.

**Comment by creker** [comment]
Looks like we're hitting this problem and it causes serious latency issues. We have 3 node baremetal cluster and occasionally get one or two cores pinged at 100% on all 3 nodes (probably due to quorum). It would be fine if it weren't affecting everything else. All other queries that I guess happen to land on the same core suffer huge latency and even timeout. As I understand it, Scylla architecture doesn't allow cores to share load for queries destined to the same shard and this problem can be considered "by design".

1. Maybe there is a way this could be fixed apart from rate limiting like this issue suggests? At least in the mean time while this issue is being worked on.
2. Does SMT affect this? Could overloaded virtual core affect queries not only on the same virtual core but also on the other one from the same physical core? Would disabling SMT at least somewhat help with this problem?
3. Is there any way to properly diagnose these kind of problems? Right now I had to basically disable queries in our app one by one to find the culprit. Once I found it the problem went away immediately. Does scylla have any tools of its own that could help find which specific queries or at least tables overload shards? It seems there's a lack of this kind of tooling. It's very hard to understand what's going on in the cluster.

**Comment by avikivity** [comment]
1. You can try catching the hot partition on the client side. It's not always feasible, it makes a lot more sense to do this on the server.
2. SMT gives you more shards, each with less capacity. Disabling SMT may help if the extra capacity is sufficient to process the hot shard's requests, but less overall throughtput. Most likely it will not help.
3. `nodetool toppartitions` can be used to detect hot keys.

In Scylla Enterprise you can use workload prioritization to protect semi-independent workloads from each other, but this isn't available in the open source version.

**Comment by avikivity** [comment]
Ideas for implementing:

1. use the `data_listener` interface we already have for `nodetool toppartitions`
2. instead of tracking partition keys, track hashed tokens. Allocate a vector hits[], and increment hits[hash(token(key))] each time we have a hit. Since the token is already available, this is cheap.
3. Every period, decay hits[*] by shifting it right by 1 (or perhaps multiplying it by a factor of 7/8 every period/3 or 15/16 every period/4 for more accuracy)

**Comment by denesb** [comment]
The data_listener is consulted after the reader which is to server the read is already created. We should apply the rate limiting as early as possible to reduce wasted work.

**Comment by avikivity** [comment]
Good point. I think we can start with how it is now to avoid rewriting the world, but later change it to wrap mutation sources:

```c++
class data_listener {
    virtual mutation_source adjust_mutation_source(mutation_source&& input) const = 0;
};
```

This gives the opportunity to react before the reader is created. We might need to add priorities so that we call the rate limiter before the toppartitions listener.

Or how about this: don't use data_listener at all. Check it in table::query. Or even coordinator-side storage_proxy, so it doesn't even reach the replica.

**Comment by denesb** [comment]
> Or how about this: don't use data_listener at all. Check it in table::query.

Yes, this should work. Alter, after we move the r_c_s in front of the cache too, maybe we can put it there. But dont' have to add it to an existing component at all costs.

> Or even coordinator-side storage_proxy, so it doesn't even reach the replica.

Won't this be a per-replica metric? I don't see how we can maintain a consistent rate across multiple nodes.

**Comment by avikivity** [comment]
> > Or even coordinator-side storage_proxy, so it doesn't even reach the replica.
> 
> Won't this be a per-replica metric? I don't see how we can maintain a consistent rate across multiple nodes.

If the client it token-aware and shard-aware, it ends up the same thing (only less work since fewer rpc messages).

If the client it not token-aware or shard-aware, then coordinator-side can under-estimate the amount of requests reaching the replicas that own the key. Perhaps we should do both: the replica-side is for ultimate safety, coordinator-side provides even more work reduction but only if the client cooperates. Let's start with replica side since that always work, and add coordinator-side protection later.

**Comment by denesb** [comment]
> > > Or even coordinator-side storage_proxy, so it doesn't even reach the replica.
> > 
> > 
> > Won't this be a per-replica metric? I don't see how we can maintain a consistent rate across multiple nodes.
> 
> If the client it token-aware and shard-aware, it ends up the same thing (only less work since fewer rpc messages).

Only true for CL_ONE messages, but even then the client will probably round-robin between replicas.

> 
> If the client it not token-aware or shard-aware, then coordinator-side can under-estimate the amount of requests reaching the replicas that own the key. Perhaps we should do both: the replica-side is for ultimate safety, coordinator-side provides even more work reduction but only if the client cooperates. Let's start with replica side since that always work, and add coordinator-side protection later.

Agree, there are a lot of questions around the possibility of moving this to a coordinator. Replica side implementation raises some questions too. What if only one replica is saturated? Will we fail the query? Won't the user be surprised that the same query sometimes works sometimes doesn't. Will the coordinator do an internal retry assuming some other replica will be able to respond? 
The coordinator can maybe maintain a heat-map of various replicas and select those that will likely be able to respond. The heat-map will have to be decayed. Note that this might take up a lot of space as we have a *log* of tokens on any replica.

**Comment by avikivity** [comment]
> > > > Or even coordinator-side storage_proxy, so it doesn't even reach the replica.
> > > 
> > > 
> > > Won't this be a per-replica metric? I don't see how we can maintain a consistent rate across multiple nodes.
> > 
> > 
> > If the client it token-aware and shard-aware, it ends up the same thing (only less work since fewer rpc messages).
> 
> Only true for CL_ONE messages, but even then the client will probably round-robin between replicas.
> 

Within a factor of CL/RF. The goal is not to have an exact boundary, but to set a limit that is an order of magnitude (or more) than what is expected for a single key, and an order of magnitude (or less) than what is expected from a shard's capacity.

(or course communicating that will be hard)

> > If the client it not token-aware or shard-aware, then coordinator-side can under-estimate the amount of requests reaching the replicas that own the key. Perhaps we should do both: the replica-side is for ultimate safety, coordinator-side provides even more work reduction but only if the client cooperates. Let's start with replica side since that always work, and add coordinator-side protection later.
> 
> Agree, there are a lot of questions around the possibility of moving this to a coordinator. Replica side implementation raises some questions too. What if only one replica is saturated? Will we fail the query? Won't the user be surprised that the same query sometimes works sometimes doesn't? Will the coordinator do an internal retry assuming some other replica will be able to respond?

See the response to the previous question. It's supposed to trigger when the key is obviously under overload (say a user querying their status 100 times per second), but the shard is obviously still healthy (100 reads/sec is less than a percent of a shard's capacity). Different use cases will tune it differently, but the exact rate shouldn't matter.

> The coordinator can maybe maintain a heat-map of various replicas and select those that will likely be able to respond. The heat-map will have to be decayed. Note that this might take up a lot of space as we have a _lot_ of tokens on any replica.

My idea involved hashing the tokens to something reasonable so there will be collisions.

**Comment by kbr-** [comment]
Observation: the toppartitions data listener is only installed for the duration of a top partitions query.

For rate limiting we could have two listeners (one of them isn't really a "listener" though):
- a listener for gathering metrics,
- a "listener" (not really since it may actually modify the data stream) for rejecting operations if they were to cross the limits.

The former we could periodically uninstall/reinstall because we don't need to gather metrics all the time, we only need a sample.
Not clear whether this would actually have a (positive) effect on performance. But it may since listeners are invoked on the hot path.

The latter could also be installed only when we actually observe that there's a problem - for example, a periodically waking up process would look at the metrics and see if they cross the limits, then install the "listener"; similarly uninstall it when (if) the situation stabilizes. This of course means that the limits could sometimes be broken but only for limited durations of time.

**Comment by haaawk** [comment]
@piodul and I have been discussing this issue and we came to the conclusion that creating a dynamic distributed rate limiting is a little bit like dynamic distributed load balancing. Meaning it's hard to get right. That led us to an idea to use a static partitioning of the limit. The way it would work is as follows:

1. Let's say that the cluster has `N` nodes and the table uses replication factor `RF`
2. Each node in the cluster can be a coordinator of a request to a partition `P` but let's distinguish two groups of coordinators: `non-replica coordinators` and `replica coordinators`. `Non-replica coordinators` are nodes that are not replicas of partition `P`. `Replica coordinators` are nodes that at the same time are replicas of partition `P`.
3. Let's assume that we want to limit the rate of operations to each partition to `L` ops/s
4. Each `non-replica coordinator` would limit the rate to `L/N` ops/s
5. Each `replica coordinator` would limit the rate to `L/RF` ops/s
6. This way we would have a `soft rate limit` for each partition of `L` ops/s and a `hard rate limit` for each partition of `2*L`.
7. The thinking is that when token-aware load balancing is used then *most* requests to partition `P` will be sent to `replica coordinators` for `P`. Those coordinators are limiting the rate to `L/RF` ops/s so the total rate will be ~`L` ops/s. It would just be a `soft rate limit` though.
8. A true limit would be achieved when token-aware load balancing is *not* used and all nodes are used as coordinators for requests to partition `P`. In such case the rate would be limited by `RF * L/RF + (N-RF) * L/N = L + (N-RF) * L/N < 2 * L`. `RF * L/RF` is the rate that goes through `replica coordinators`. `(N - RF) * L/N` is the rate that goes through `non-replica coordinators`
9. There's an optional optimisation that's based on the fact that `replica coordinators` can monitor the load on their replica sides and based on that adjust their rate limit to be anything in range `[L/N, L/RF]` dynamically.

This is a very simple engineering approach to the problem which is far from perfect but it has a benefit that it should be quite easy to implement it right while the effectiveness of the solution is still quite good. 

The assumption that supported this thinking was that rate limit for each partition will probably be a small fraction of the node capacity because a node hosts multiple partitions and we don't want any hot partition to consume the whole node or even a big fraction of it. 

@avikivity @dorlaor would such a solution be acceptable?

**Comment by dorlaor** [comment]
A soft limit is certainly the right way. The main motivation for rate limit
isn't to charge accurately, it's
usually to block excessive accesses that risk the cluster or the shard,
something
which is far higher than the average operations per partition. In such a
case, it will be enough to have
a cap per shard per partition. We can enforce it on all shards in the
coordination phase (whether they
own the data or not)

**Comment by avikivity** [comment]
> @piodul and I have been discussing this issue and we came to the conclusion that creating a dynamic distributed rate limiting is a little bit like dynamic distributed load balancing. Meaning it's hard to get right. That led us to an idea to use a static partitioning of the limit. The way it would work is as follows:
> 
> 1. Let's say that the cluster has `N` nodes and the table uses replication factor `RF`
> 2. Each node in the cluster can be a coordinator of a request to a partition `P` but let's distinguish two groups of coordinators: `non-replica coordinators` and `replica coordinators`. `Non-replica coordinators` are nodes that are not replicas of partition `P`. `Replica coordinators` are nodes that at the same time are replicas of partition `P`.
> 3. Let's assume that we want to limit the rate of operations to each partition to `L` ops/s
> 4. Each `non-replica coordinator` would limit the rate to `L/N` ops/s
> 5. Each `replica coordinator` would limit the rate to `L/RF` ops/s
> 6. This way we would have a `soft rate limit` for each partition of `L` ops/s and a `hard rate limit` for each partition of `2*L`.
> 7. The thinking is that when token-aware load balancing is used then _most_ requests to partition `P` will be sent to `replica coordinators` for `P`. Those coordinators are limiting the rate to `L/RF` ops/s so the total rate will be ~`L` ops/s. It would just be a `soft rate limit` though.
> 8. A true limit would be achieved when token-aware load balancing is _not_ used and all nodes are used as coordinators for requests to partition `P`. In such case the rate would be limited by `RF * L/RF + (N-RF) * L/N = L + (N-RF) * L/N < 2 * L`. `RF * L/RF` is the rate that goes through `replica coordinators`. `(N - RF) * L/N` is the rate that goes through `non-replica coordinators`
> 9. There's an optional optimisation that's based on the fact that `replica coordinators` can monitor the load on their replica sides and based on that adjust their rate limit to be anything in range `[L/N, L/RF]` dynamically.

We have too many coordinators - consider 20 nodes and 50 shards/node, that's 1000 coordinators. If you want to limit 30 keys/sec in a 10 second period, that's 0.3 keys/sec/coordinator shard.

It's easiest to do this on the replica. It means more work is wasted, but it's probably still okay. I guess the replica can teach the coordinator - if the replica responds with {deny, 30sec} the coordinator can deny that key for 30 seconds.

> 
> This is a very simple engineering approach to the problem which is far from perfect but it has a benefit that it should be quite easy to implement it right while the effectiveness of the solution is still quite good.
> 
> The assumption that supported this thinking was that rate limit for each partition will probably be a small fraction of the node capacity because a node hosts multiple partitions and we don't want any hot partition to consume the whole node or even a big fraction of it.
> 
> @avikivity @dorlaor @tzach would such a solution be acceptable?

**Comment by dorlaor** [comment]
On Wed, Nov 3, 2021 at 3:26 PM Avi Kivity ***@***.***> wrote:

> @piodul <https://github.com/piodul> and I have been discussing this issue
> and we came to the conclusion that creating a dynamic distributed rate
> limiting is a little bit like dynamic distributed load balancing. Meaning
> it's hard to get right. That led us to an idea to use a static partitioning
> of the limit. The way it would work is as follows:
>
>    1. Let's say that the cluster has N nodes and the table uses
>    replication factor RF
>    2. Each node in the cluster can be a coordinator of a request to a
>    partition P but let's distinguish two groups of coordinators: non-replica
>    coordinators and replica coordinators. Non-replica coordinators are
>    nodes that are not replicas of partition P. Replica coordinators are
>    nodes that at the same time are replicas of partition P.
>    3. Let's assume that we want to limit the rate of operations to each
>    partition to L ops/s
>    4. Each non-replica coordinator would limit the rate to L/N ops/s
>    5. Each replica coordinator would limit the rate to L/RF ops/s
>    6. This way we would have a soft rate limit for each partition of L
>    ops/s and a hard rate limit for each partition of 2*L.
>    7. The thinking is that when token-aware load balancing is used then
>    *most* requests to partition P will be sent to replica coordinators
>    for P. Those coordinators are limiting the rate to L/RF ops/s so the
>    total rate will be ~L ops/s. It would just be a soft rate limit though.
>    8. A true limit would be achieved when token-aware load balancing is
>    *not* used and all nodes are used as coordinators for requests to
>    partition P. In such case the rate would be limited by RF * L/RF +
>    (N-RF) * L/N = L + (N-RF) * L/N < 2 * L. RF * L/RF is the rate that
>    goes through replica coordinators. (N - RF) * L/N is the rate that
>    goes through non-replica coordinators
>    9. There's an optional optimisation that's based on the fact that replica
>    coordinators can monitor the load on their replica sides and based on
>    that adjust their rate limit to be anything in range [L/N, L/RF]
>    dynamically.
>
> We have too many coordinators - consider 20 nodes and 50 shards/node,
> that's 1000 coordinators. If you want to limit 30 keys/sec in a 10 second
> period, that's 0.3 keys/sec/coordinator shard.
>
> It's easiest to do this on the replica. It means more work is wasted, but
> it's probably still okay. I guess the replica can teach the coordinator -
> if the replica responds with {deny, 30sec} the coordinator can deny that
> key for 30 seconds.
>

But there is no point having a cap of 30 key/sec. Also, most times the
client is topology aware.
A normal cap to block a spam attack or any other unpredictable load, like
the dress that broke
the internet will require a high cap - 1000 op/key/sec or even 10k. In such
a case, we can
enforce it on the coordinator data owner. We can enforce all non-owner
replicas with a cap too.
They all don't need to be exact, to have a cap of T (while T is 1k..10k
ops) we can set a T/2
per replica (at the coordinator entrance) and T/2 on the non-owners. No
need to be 100% accurate,
so an actual enforcement of 1.5T is fine too

**Comment by haaawk** [comment]
I share @dorlaor's opinion that 30 key/sec sounds unrealistic. I would also expect something in the range of (tens of) thousands of ops/s

**Comment by avikivity** [comment]
The capacity of a shard is thousands to tens of thousands of ops/sec. So if the limit is thousand or tens of thousands of ops/sec, you're leaving nothing for other partitions.

You can also look at it from another direction. If the partition represents a user's interactions with the system, then anything above 30 ops/sec is certainly a bot.

**Comment by tzach** [comment]
For reference
The assumption in Scylla Calc is sustainable 15K operations per vCPU 
https://github.com/scylladb/scylla-price-calculator#model-assumptions

**Comment by haaawk** [comment]
Fair enough. I don't think interaction created by bots is forbidden though, is it?

**Comment by haaawk** [comment]
So can we assume the limit will be in range of tens or hundreds of ops/s? That can influence design.

**Comment by avikivity** [comment]
Best to avoid such assumptions unless critical. We don't know how this will be used.

**Comment by haaawk** [comment]
I know it's best to keep the code as general as possible but also having some assumptions can lead to a simpler, better performing code that's delivered faster. We can always change the implementation if it turns out we need limit of thousands of ops/s. Your own argument stands really strongly here if we make the limit 1k then a shard will be able to keep only 15 partitions. Is that enough?

As a middle ground, maybe assumption that the limit is < 10k would be possible? It's highly improbable that there's ~1 partition per shard and then limiting makes little sense anyway.

*I'm basing those calculations on @tzach 15k ops pre vCPU note
**Also this assumption would make sense only if it turns out it simplifies code and makes it more performant. Otherwise it makes no sense.

**Comment by slivne** [comment]
as a reference aerospike has a hotkey concept as well https://discuss.aerospike.com/t/hot-key-error-code-14/986 I am not sure its for the same usage but I think some users refereed to this in comparison in the past.

**Comment by haaawk** [comment]
But going back to the static partitioning of the limit. Even in your 30ops/s example it would work as follows:
1. each replica coordinator would get 10ops/s limit assuming RF=3
2. each non-replica coordinator would get 1ops/s limit by rounding up

For a properly configured driver that would still mean ~30ops/s limit when a token aware load balancing is used.

**Comment by avikivity** [comment]
> I know it's best to keep the code as general as possible but also having some assumptions can lead to a simpler, better performing code that's delivered faster. We can always change the implementation if it turns out we need limit of thousands of ops/s. Your own argument stands really strongly here if we make the limit 1k then a shard will be able to keep only 15 partitions. Is that enough?

It depends on the distribution of hot partitions. I really can't say much as we don't have experience here.

If you show a data structure that behaves better with the assumptions, we can consider it. I have ideas about a data structure that is (mostly) rate independent.


> 
> As a middle ground, maybe assumption that the limit is < 10k would be possible? It's highly improbable that there's ~1 partition per shard and then limiting makes little sense anyway.


That's for sure.

> 
> *I'm basing those calculations on @tzach 15k ops pre vCPU note **Also this assumption would make sense only if it turns out it simplifies code and makes it more performant. Otherwise it makes no sense.

**Comment by avikivity** [comment]
> But going back to the static partitioning of the limit. Even in your (@avikivity) 30ops/s example it would work as follows:
> 
> 1. each replica coordinator would get 10ops/s limit assuming RF=3
> 2. each non-replica coordinator would get 1ops/s limit by rounding up

per shard, the numbers are much worse. Do we want cross shard coordination for this?

> 
> For a properly configured driver that would still mean ~30ops/s limit when a token aware load balancing is used.

and shard aware!

> For a misconfigured driver it would be bigger but still limited. The thing is that driver that's not configured to cooperate will just retry instantly and the whole limiting will have little effect.
